File: transektcount/src/main/java/com/wmstein/transektcount/PermissionsDialogFragment.java
package com.wmstein.transektcount;

import android.Manifest;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.provider.Settings;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.support.v7.app.AlertDialog;

/**
 * PermissionsDialogFragment provides the permission handling, which is
 * necessary since Android Marshmallow (M)
 * 
 * Original version from RuntimePermissionsExample-master created by tylerjroach on 8/31/16,
 * licensed under the MIT License.
 * 
 * Adopted for TourCount by wistein on 2018-06-20,
 * last edited on 2018-08-03.
 */

public class PermissionsDialogFragment extends DialogFragment
{
    private final int PERMISSION_REQUEST_CODE = 101;

    private Context context;
    private PermissionsGrantedCallback listener;

    private boolean shouldResolve;
    private boolean shouldRetry;
    private boolean externalGrantNeeded;

    public static PermissionsDialogFragment newInstance()
    {
        return new PermissionsDialogFragment();
    }

    public PermissionsDialogFragment()
    {
    }

    @Override
    public void onAttach(Context context)
    {
        super.onAttach(context);
        this.context = context;
        if (context instanceof PermissionsGrantedCallback)
        {
            listener = (PermissionsGrantedCallback) context;
        }
    }

    @Override
    public void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        setStyle(STYLE_NO_TITLE, R.style.PermissionsDialogFragmentStyle);
        setCancelable(false);
        requestNecessaryPermissions();
    }

    @Override
    public void onResume()
    {
        super.onResume();
        if (shouldResolve)
        {
            if (externalGrantNeeded)
            {
                showAppSettingsDialog();
            }

            else if (shouldRetry)
            {
                showRetryDialog();
            }
           
            else
            {
                //permissions have been accepted
                if (listener != null)
                {
                    listener.permissionCaptureFragment();
                    dismiss();
                }
            }
        }
    }

    @Override
    public void onDetach()
    {
        super.onDetach();
        context = null;
        listener = null;
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String permissions[], @NonNull int[] grantResults)
    {
        shouldResolve = true;
        shouldRetry = false;

        for (int i = 0; i < permissions.length; i++)
        {
            String permission = permissions[i];
            int grantResult = grantResults[i];

            if (!shouldShowRequestPermissionRationale(permission) && grantResult != PackageManager.PERMISSION_GRANTED)
            {
                externalGrantNeeded = true;
                return;
            }
            else if (grantResult != PackageManager.PERMISSION_GRANTED)
            {
                shouldRetry = true;
                return;
            }
        }
    }

    private void requestNecessaryPermissions()
    {
        requestPermissions(new String[]{
            Manifest.permission.WRITE_EXTERNAL_STORAGE}, 
            PERMISSION_REQUEST_CODE);
    }

    private void showAppSettingsDialog()
    {
        new AlertDialog.Builder(context)
            .setTitle(getString(R.string.perm_required))
            .setMessage(getString(R.string.perm_hint) + " " + getString(R.string.perm_hint1))
            .setPositiveButton(getString(R.string.app_settings), new DialogInterface.OnClickListener()
            {
                @Override
                public void onClick(DialogInterface dialogInterface, int i)
                {
                    Intent intent = new Intent();
                    intent.setAction(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
                    Uri uri = Uri.fromParts("package", context.getApplicationContext().getPackageName(), null);
                    intent.setData(uri);
                    context.startActivity(intent);
                    dismiss();
                }
            })
            .setNegativeButton(getString(R.string.cancel), new DialogInterface.OnClickListener()
            {
                @Override
                public void onClick(DialogInterface dialogInterface, int i)
                {
                    dismiss();
                }
            }).create().show();
    }


    private void showRetryDialog()
    {
        new AlertDialog.Builder(context)
            .setTitle(getString(R.string.perm_declined))
            .setMessage(getString(R.string.perm_hint))
            .setPositiveButton(getString(R.string.retry), new DialogInterface.OnClickListener()
            {
                @Override
                public void onClick(DialogInterface dialogInterface, int i)
                {
                    requestNecessaryPermissions();
                }
            })
            .setNegativeButton(getString(R.string.cancel), new DialogInterface.OnClickListener()
            {
                @Override
                public void onClick(DialogInterface dialogInterface, int i)
                {
                    dismiss();
                }
            }).create().show();
    }


    public interface PermissionsGrantedCallback
    {
        void permissionCaptureFragment();
    }
}

File: transektcount/src/main/java/com/wmstein/transektcount/PermissionsDialogFragment.kt
package com.wmstein.transektcount

import android.Manifest
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.util.Log
import androidx.activity.result.contract.ActivityResultContracts
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.DialogFragment

/**
 * PermissionsDialogFragment provides the permission handling,
 * which is necessary since Android Marshmallow (M)
 *
 * Created in Kotlin on 2023-05-26,
 * last edited on 2025-02-21
 */
class PermissionsDialogFragment : DialogFragment() {
    private var context: Context? = null
    private var shouldResolve = false
    private var externalGrantNeeded = false
    private var externalGrant30Needed = false

    override fun onAttach(context: Context) {
        super.onAttach(context)

        this.context = context
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setStyle(STYLE_NO_TITLE, R.style.PermissionsDialogFragmentStyle)
        isCancelable = false

        // Check for given storage permission
        requestStoragePermissions()
    }

    // Request storage permission
    private fun requestStoragePermissions() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            // launch permission request dialog (Android >=11)
            val permission = Manifest.permission.MANAGE_EXTERNAL_STORAGE
            permissionLauncherStorage.launch(permission)
        } else {
            // Android < 11
            val permission = Manifest.permission.WRITE_EXTERNAL_STORAGE
            permissionLauncherStorage.launch(permission)
        }
    }

    // Request single permissions in system settings app
    private val permissionLauncherStorage = registerForActivityResult(
        ActivityResultContracts.RequestPermission()
    )
    { isGranted ->
        shouldResolve = true
        if (MyDebug.DLOG) Log.d(TAG, "65, onActivityResult: isGranted: $isGranted")

        if (isGranted) {
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) {
                externalGrantNeeded = false
            } else {
                externalGrant30Needed = false
            }
            dismiss()
        } else {
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) {
                externalGrantNeeded = true
            } else {
                externalGrant30Needed = true
            }
            if (MyDebug.DLOG) Log.d(TAG, "79, onActivityResult: Permission denied...")
        }
    }

    override fun onResume() {
        super.onResume()

        if (shouldResolve) {
            if (externalGrantNeeded) {
                if (MyDebug.DLOG) Log.i(TAG, "90, ask storage")
                showAppSettingsStorageDialog()
            }
            if (externalGrant30Needed) {
                // >= API 30
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                    if (MyDebug.DLOG) Log.i(TAG, "95, ask manage storage")
                    showAppSettingsManageStorageDialog()
                }
            }
        }
    }

    // Query missing  external storage permissions for API <30
    private fun showAppSettingsStorageDialog() {
        AlertDialog.Builder(requireContext())
            .setTitle(getString(R.string.dialog_storage_title))
            .setMessage(getString(R.string.dialog_storage_message))
            .setPositiveButton(getString(R.string.app_settings))
            { _: DialogInterface?, _: Int ->
                val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
                val uri = Uri.fromParts("package", "com.wmstein.transektcount", null)
                intent.data = uri
                requireContext().startActivity(intent)
                dismiss()
            }
            .setNegativeButton(getString(R.string.cancel))
            { _: DialogInterface?, _: Int ->
                dismiss()
            }
            .create().show()
    }

    // Query missing manage storage permission for API >=30
    @RequiresApi(Build.VERSION_CODES.R)
    private fun showAppSettingsManageStorageDialog() {
        val intent = Intent(Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION)
        val uri = Uri.fromParts("package", "com.wmstein.transektcount", null)
        intent.data = uri
        startActivity(intent)
        dismiss()
    }

    override fun onDetach() {
        super.onDetach()

        context = null
    }

    companion object {
        private const val TAG = "PermStorFragm"

        @JvmStatic
        fun newInstance(): PermissionsDialogFragment {
            return PermissionsDialogFragment()
        }
    }

}
